packages/api-server/src/error.ts

```diff
@@ -0,0 +1,160 @@
+//  Error code from JSON-RPC 2.0 spec
+//  reference: http://www.jsonrpc.org/specification#error_object
+
+import { JSONRPCError } from "jayson";
+import { COMPATIBLE_DOCS_URL } from "./methods/constant";
+
+export interface ErrorKind {
+  code: number;
+  message: string;
+}
+
+export class AppError extends Error implements JSONRPCError {
+  code: number;
+  data?: object;
+
+  constructor(kind: ErrorKind, data?: object) {
+    super(kind.message);
+    this.name = kind.message;
+    this.code = kind.code;
+    this.data = data;
+  }
+
+  addContext(context: Object): AppError {
+    this.data = {
+      ...this.data,
+      ...context,
+    };
+    return this;
+  }
+}
+
+const DEFAULT_ERROR_CODE: number = -32000;
+
```

Comment:
```suggestion
export const INVALID_PARAMS_CODE = -32602;
```
pls extracts all error code to const here.

```sh
export const ERRORS = Object.freeze({
  INVALID_PARAMETER: {
    code: INVALID_PARAMS_CODE,
    message: "invalid parameter",
  } as ErrorKind,
```

Reply:
Actually, I have tried "freeze". However, I gave up that time since my IDE cannot autocomplete the frozen items :(. I have no idea why.
I have no strong opinions about it.
BTW, is "freeze" a kind of best practice?

Reply:
[immutable.js](https://immutable-js.com/) or type with read-only properties?

```sh
type ErrorKind = {
  readonly code: number;
  readonly message: string;
};
```

Reply:
> [immutable.js](https://immutable-js.com/) or type with read-only properties?
> 
> ```shell
> type ErrorKind = {
>   readonly code: number;
>   readonly message: string;
> };
> ```

cool!